### DISCUSSION TITLE
How to use only shell to execute few commands

### DISCUSSION THREAD

**Discussion by suryadevelope** [termux/termux-app#3555] [General]
Hey i would like to create a app with small features of termux only with the environment and need to install few packages like python, nodejs etc but dont need any designer and other functionalities of the termux 

i was confused where to start and how to proceed 

can any one help me to to achieve it please
i have gone through  termux wiki also and tried some part of it 

as of my observations the environment is created in my app but i cannot install a new package and i am not sure about the procedure i followed i can provide my code 

```
  setLogConfig(context);

        Logger.logDebug("Starting Application");

        TermuxCrashUtils.setCrashHandler(this);


        TermuxSharedProperties properties = new TermuxSharedProperties(this);

        properties.dumpPropertiesToLog();
        properties.getProperties(true);

        properties.loadTermuxPropertiesFromDisk();

//        mPreferences = TermuxAppSharedPreferences.build(this);
//        if (mPreferences == null) {
//            // An AlertDialog should have shown to kill the app, so we don't continue running activity code
//            Toast.makeText(context, "close the app", Toast.LENGTH_SHORT).show();
//            return;
//        }



        TermuxShellEnvironmentClient termuxShellEnvironmentClient = new TermuxShellEnvironmentClient();
        termuxShellEnvironmentClient.buildEnvironment(this, false,null);





        // Check and create termux files directory. If failed to access it like in case of secondary
        // user or external sd card installation, then don't run files directory related code
        Error error = isTermuxFilesDirectoryAccessible(this, true, true);
        boolean isTermuxFilesDirectoryAccessible = error == null;
        if (isTermuxFilesDirectoryAccessible) {
            Log.e("surya","Termux files directory is accessible");
            Logger.logInfo(LOG_TAG, "Termux files directory is accessible");

            error = isAppsTermuxAppDirectoryAccessible(true, true);
            if (error != null) {
                Logger.logErrorExtended(LOG_TAG, "Create apps/termux-app directory failed\n" + error);
                return;
            }

            // Setup termux-am-socket server
        } else {
            Logger.logErrorExtended(LOG_TAG, "Termux files directory is not accessible\n" + error);
        }

        Log.e("TermuxShellEnvironmentClient",termuxShellEnvironmentClient.getDefaultWorkingDirectoryPath());



        TermuxSession.TermuxSessionClient termuxSessionClient = new TermuxSession.TermuxSessionClient() {
            @Override
            public void onTermuxSessionExited(TermuxSession termuxSession) {
                Log.e("suryasessiondata : ", String.valueOf(termuxSession.getExecutionCommand().resultData));
                Log.e("suryasessionstdout : ", String.valueOf(termuxSession.getExecutionCommand().resultData.stdout));
                Log.e("suryasessionstderror : ", String.valueOf(termuxSession.getExecutionCommand().resultData.stderr));
                Log.e("suryasessionerrlist : ", String.valueOf(termuxSession.getExecutionCommand().resultData.errorsList));
                Log.e("suryasessioncode : ", String.valueOf(termuxSession.getExecutionCommand().resultData.exitCode));
            }
        };

//        TerminalView terminalView = findViewById(R.id.terminal_view);


        TerminalSessionClient terminalSessionClient = new TerminalSessionClient() {
            @Override
            public void onTextChanged(TerminalSession changedSession) {
                Log.e("suryaj : ", String.valueOf(changedSession));
            }

            @Override
            public void onTitleChanged(TerminalSession changedSession) {
                Log.e("suryatchange : ", String.valueOf(changedSession));

            }

            @Override
            public void onSessionFinished(TerminalSession finishedSession) {
                Log.e("suryaf : ", String.valueOf(finishedSession.getCwd()));

            }

            @Override
            public void onClipboardText(TerminalSession session, String text) {
                Log.e("suryaclip : ", String.valueOf(text));


            }

            @Override
            public void onBell(TerminalSession session) {
                Log.e("suryabell : ", String.valueOf(session));

            }

            @Override
            public void onColorsChanged(TerminalSession session) {

            }

            @Override
            public void onTerminalCursorStateChange(boolean state) {

            }

            @Override
            public Integer getTerminalCursorStyle() {
                return null;
            }

            @Override
            public void logError(String tag, String message) {
                Log.e("suryae : ", String.valueOf(message));

            }

            @Override
            public void logWarn(String tag, String message) {
                Log.e("suryaw : ", String.valueOf(message));

            }

            @Override
            public void logInfo(String tag, String message) {
                Log.e("suryaI : ", String.valueOf(message));

            }

            @Override
            public void logDebug(String tag, String message) {
                Log.e("suryad : ", String.valueOf(message));

            }

            @Override
            public void logVerbose(String tag, String message) {
                Log.e("suryav : ", String.valueOf(message));

            }

            @Override
            public void logStackTraceWithMessage(String tag, String message, Exception e) {

            }

            @Override
            public void logStackTrace(String tag, Exception e) {

            }
        };
        TermuxShellUtils.buildEnvironment(this, true,null);


        TermuxSession termuxSession =TermuxSession.execute(this,new ExecutionCommand(101,null,null,"ls", "home",false,false), terminalSessionClient, termuxSessionClient, termuxShellEnvironmentClient, "surya exec",true);
        Log.e("kjhgf", String.valueOf(!termuxSession.getExecutionCommand().setState(ExecutionCommand.ExecutionState.EXECUTING)));
        termuxSessionClient.onTermuxSessionExited(termuxSession);

        Log.e("suryaee", String.valueOf(termuxSession.getTerminalSession()));
        Log.e("suryaee", String.valueOf(termuxSession.getExecutionCommand().getCommandDescriptionLogString()));

```

all the code is error free and app builded successfully

**Comment by agnostic-apollo** [comment]
>as of my observations the environment is created in my app but i cannot install a new package and i am not sure about the procedure i followed i can provide my code

Random code snippets are mostly useless to understand what you have done, termux is quite complex and requires activity and service and other stuff to function properly. You have also not posted any errors for failures, we can't guess what's is wrong.

You also need to compile all the packages yourself for your app package. I suggest you read https://github.com/termux/termux-app#forking

There are also `targetSdkVersion ` `>= 28` restrictions on Android. Check https://github.com/termux/termux-packages/wiki/Termux-and-Android-10

There are no examples for extracting a terminal out of termux currently. What parts of termux you will require really depends on what functionalities you want your terminal to support, like extra keys, multiple sessions, theming. From termux-app app module you mainly need parts of `TermuxApplication`, `TermuxActivity`, `TermuxService`, `terminal` package. Some base implementations are provided. If you don't specifically require a terminal and just want to run commands, use `AppShell` instead. `TaskerAppFactory` also uses it and its design is compatible with `targetSdkVersion ` `>= 28`.

https://github.com/termux/termux-app/tree/master/app/src/main/java/com/termux/app/terminal

https://github.com/agnostic-apollo/TaskerAppFactory

**Comment by v-python** [comment]
Is AppShell a standalone piece of termux, or part of something else (LS Central has an AppShell on Google Play)?
I'd been using termux from Google Play mostly as an environment to run a single Python script,  Worked great until I upgraded to a Pixel 7, and now the Google Play version doesn't work, and termux seemed like it was big enough to install without also having to install fdroid to get a version of termux to run. Of course, what I'd really like is a standalone Python app, that could be invoked by clicking a script file shortcut.  With configuration, I sort of could make termux be that, but there is even more overhead now, with needing fdroid + termux to get Python.

**Reply by agnostic-apollo** [reply]
`AppShell` is the java class that is a wrapper for `Runtime.exec()` that termux uses to run background shell commands, that can also be used by other apps if they import the `termux-shared` library. `TaskerAppFactory` repo just ports the class for its own usage so that it doesn't need to add `termuxs-shared` as a dependency. The caller still needs to set up the environment to run the commands, there are separate classes for that in the library. For python, etc you would need to compile a bootstrap including the required packages for your app package and export environment accordingly.

https://github.com/termux/termux-app/blob/55cdef01e7b2c0c1c7946a95e716098bf441256c/termux-shared/src/main/java/com/termux/shared/shell/command/runner/app/AppShell.java#L37

https://github.com/agnostic-apollo/TaskerAppFactory/blob/dbf319ffdb39e99ceb43cdcddc0799a3698be840/app/src/main/java/net/dinglisch/android/appfactory/utils/shell/runner/app/AppShell.java#L38

https://developer.android.com/reference/java/lang/Runtime#exec(java.lang.String[],%20java.lang.String[],%20java.io.File)

https://github.com/termux/termux-packages/wiki/For-maintainers#build-bootstrap-archives

>but there is even more overhead now, with needing fdroid + termux to get Python.

Playstore is not supported anymore. For F-Droid, you don't need to install their app, you can download termux apk from their site and install directly. Or install github builds.

https://github.com/termux/termux-app#installation

https://github.com/termux/termux-app#google-play-store-deprecated
